### ISSUE TITLE
Batch downloads and uploads

### ISSUE THREAD

**Issue by dri94** [Storj/android-libstorj#11]
Currently we cannot do batch downloads without getting an error if trying to download too many at once (Have to use a loop to download currently). A method for batch downloads that takes a list of files to download, and returns a list of the downloaded files should be created.

**Comment by kaloyan-raev** [comment]
What's the exact error you get?

**Comment by dri94** [comment]
Its a bunch of failed to initialize Storj environment errors. Class looks like this.
```
class A {
private Storj storj;
public ExampleClass() {
storj = Storj.getInstance();
}

public void downloadWithStorj() {
storj.downloadFile(blah, blah, blah);
}
}
```
```
class B {
public class B() {
A a = new A();
}
public void someMethod() {
for (int i = 0; i < 1000; i++) {
a.downloadWithStorj();
}
}
}
```
So i am using one storj instance. And calling downloadFile multiples times with a list of storj files. The first couple files download, after that i get numerous "Failed to initialize Storj environment" errors which will eventually freeze the application

**Comment by kaloyan-raev** [comment]
"Failed to initialize Storj environment" most probably means that the native library runs out of memory due to the many parallel downloads. I will need to do some additional analyse to understand what's going on and what would be the best way to introduce batch downloads.

Until then I recommend to queue the downloads and start the next only after the `onComplete` event is received for the previous one.

**Comment by dri94** [comment]
Okay will do

**Comment by adevezin** [comment]
In this thread, would you rater me add Batch uploads to this ticket, or create a separate issue for that?

**Comment by kaloyan-raev** [comment]
I updated the title.

**Comment by adevezin** [comment]
Hey, just checking in for any updates?

**Comment by kaloyan-raev** [comment]
Not yet. It's in my todo list, but not at the very top.

**Comment by kaloyan-raev** [comment]
Batch operations are introduced in version 0.8 of java-libstorj, which is included in version 0.7 of android-libstorj.

Check the Javadoc for details of the new API: http://javadoc.io/doc/io.storj/libstorj-java/0.8
